portal/portal-impl/impl/src/java/org/sakaiproject/portal/charon/site/PortalSiteHelperImpl.java

```diff
@@ -315,25 +323,57 @@ private List<String> getPinnedSiteIds() {
 		.getPropertyList("order");
 	}
 
-	private List<String> getRecentSiteIds() {
+	private List<String> getRecentSiteIds(Site currentSite) {
 
+		if ( currentSite == null || currentSite.getId() == null ) return Collections.emptyList();
+		String currentSiteId = currentSite.getId();
 		Session session = SessionManager.getCurrentSession();
+		if ( session == null ) return Collections.emptyList();
 		String userId = session.getUserId();
-		int maxRecentSites = ServerConfigurationService.getInt("portalPath", 3);
+		if ( userId == null ) return Collections.emptyList();
 
-		//Get list of last visited site ids ingoring users home site
-		String sql = String.format("SELECT DISTINCT se.CONTEXT FROM SAKAI_EVENT se, SAKAI_SESSION ss " +
-		"WHERE se.EVENT = 'pres.begin' AND se.SESSION_ID = ss.SESSION_ID " +
-		"AND ss.SESSION_USER = '%s' AND se.CONTEXT != '~%s' " +
-		"AND se.context != '!error' ORDER BY se.EVENT_DATE DESC LIMIT %d;", userId, userId, maxRecentSites);
+		int maxRecentSites = ServerConfigurationService.getInt("portalPath", 3);
 
-		List<String> siteIds = sqlService.dbRead(sql);
+		if ( currentSiteId.startsWith("~") || currentSiteId.startsWith("!") ) {
+			Preferences prefs = PreferencesService.getPreferences(session.getUserId());
+			ResourceProperties props = prefs.getProperties(org.sakaiproject.user.api.PreferencesService.SITENAV_PREFS_KEY);
 
-		if (siteIds != null) {
-			return siteIds;
+			String recentStr = props.getProperty(PROP_RECENT_SITES);
+			ArrayList<String> recents = new ArrayList<String>(StringUtils.isBlank(recentStr) ? Collections.emptyList() : Arrays.asList(recentStr.split("::")));
```

Comment:
```suggestion
			List<String> recents = StringUtils.isBlank(recentStr) ? Collections.<String>emptyList() : Arrays.asList(recentStr.split("::")));
```

Reply:
With your change, it throws org.sakaiproject.portal.api.PortalHandlerException: java.lang.UnsupportedOperationException: remove

This needs to be forced to make an ArrrayList because so many methods are optional in List